sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala

```diff
@@ -477,5 +486,82 @@ object StateStore extends Logging {
       None
     }
   }
-}
 
+  private class StateSchemaCompatibilityChecker(
+      providerId: StateStoreProviderId,
+      hadoopConf: Configuration) {
+
+    private val storeCpLocation = providerId.storeId.storeCheckpointLocation()
+    private val fm = CheckpointFileManager.create(storeCpLocation, hadoopConf)
+    private val schemaFileLocation = schemaFile(storeCpLocation)
+
+    fm.mkdirs(schemaFileLocation.getParent)
+
+    def check(keySchema: StructType, valueSchema: StructType): Unit = {
+      if (fm.exists(schemaFileLocation)) {
+        logDebug(s"Schema file for provider $providerId exists. Comparing with provided schema.")
+        val (storedKeySchema, storedValueSchema) = readSchemaFile()
+
+        def typesEq(schema1: StructType, schema2: StructType): Boolean = {
+          val fieldToType: StructField => (DataType, Boolean) = f => (f.dataType, f.nullable)
+          (schema1.length == schema2.length) &&
+            schema1.map(fieldToType).equals(schema2.map(fieldToType))
+        }
+
+        def namesEq(schema1: StructType, schema2: StructType): Boolean = {
+          schema1.fieldNames.toList.equals(schema2.fieldNames.toList)
+        }
+
+        val errorMsg = "Provided schema doesn't match to the schema for existing state! " +
+          "Please note that Spark allow difference of field name: check count of fields " +
+          "and data type of each field.\n" +
+          s"- provided schema: key $keySchema value $valueSchema\n" +
+          s"- existing schema: key $storedKeySchema value $storedValueSchema\n"
+
+        if (!typesEq(keySchema, storedKeySchema) || !typesEq(valueSchema, storedValueSchema)) {
+          logError(errorMsg)
+          throw new IllegalStateException(errorMsg)
+        } else if (!namesEq(keySchema, storedKeySchema) ||
+          !namesEq(valueSchema, storedValueSchema)) {
+          logInfo("Detected the change of field name, will overwrite schema file to new.")
```

Comment:
Just for the sake of my understanding which use-case is this from user perspective?

Reply:
End users can change the column of aggregated column like `sum("field").as("sum_val")` to `sum("field").as("sum_value")` which doesn't change semantic. One counter case is swapping columns for min and max.